# [makemore](https://www.youtube.com/watch?v=PaCmpygFfXo&list=PLAqhIrjkxbuWI23v9cThsA9GvCAUhRvKZ&index=2)

We start by opening up the `names.txt` file in the `data/` folder and look at some attributes.

In [3]:
words = open("data/names.txt", "r").read().splitlines()

In [4]:
words[:10]

['emma',
 'olivia',
 'ava',
 'isabella',
 'sophia',
 'charlotte',
 'mia',
 'amelia',
 'harper',
 'evelyn']

In [5]:
len(words)

32033

In [6]:
min(len(w) for w in words)

2

In [7]:
max(len(w) for w in words)

15

Let's consider a name from the list in terms of what we want our model to do. If we examine the first name, `"emma"`, our model will learn from this word that it's very likely that the letter `"e"` will start a sequence of characters in a word. Similarly, the character `"m"` is likely to come after `"e"`, the character `"m"` is likely to come after `"em"`, and the character `"a"` is likely to come after `"emm"`. Finally, after the sequence `"emma"`, the word is likely to end.

Because we have so many words, `32033` from Cell `[5]`, we have a lot of structure to train our model on.

### Bigram language model

We're going to start by learning how to implement a **bigram** language model. In a bigram language model, we're always working with just **two characters at a time**. In other words, we're looking at the input character we're given and we attempt to predict the next character in the sequence.

We'll begin by looking at bigrams in our dataset.

In [10]:
for w in words[:3]:
  for ch1, ch2 in zip(w, w[1:]):
    print(ch1, ch2)
  print()

e m
m m
m a

o l
l i
i v
v i
i a

a v
v a

